process/HOPE-3--Dependency-Injection.md

```diff
@@ -0,0 +1,169 @@
+# HOPE 3 -- Dependency Injection
+
+|             |                                             |
+| ------------| ------------------------------------------- |
+| HOPE:       | 3                                           |
+| Title:      | Dependency Injection                        |
+| Author(s):  | Timothy Crosley <timothy.crosley@gmail.com> |
+| Status:     | Proposed                                    |
+| Type:       | Standard Track                              |
+| Created:    | 25-May-2019                                 |
+| Updated:    | 27-May-2019                                 |
+
+## Abstract
+
+This HOPE proposes adding a robust and universally available dependency injection system to Hug.
+This system would be built with the intention of replacing Hug's current `directive` system, one of the first Python microservice dependency injection systems, solving the shortcomings identified upon extensive usage:
+
+- Directives aren't clearly separated from type annotations.
+- It is impossible to have both a directive and type hint for a single parameter.
+- It isn't easy to swap out directives for full HTTP stack testing (through direct calls make it easy to substitute out).
+- Directives aren't nestable.
+
+## Proposed Dependency Injection System
+
+This HOPE proposes a new dependency injection system with an aim for clarity and reuse.
+The API of the new system would consist of a `provides` decorator for defining available dependencies, and a `requires` function for using them:
+
+```python
+@provides(name: str = None, api: hug.API = None, override: bool = False)
+
+def requires(name_or_function: Union[str, Callable], *args, **kwargs) -> Any
+```
+
+These two calls are meant to replace the old `Directive` system:
+
+```python
+@directive(apply_globally: bool = False, api: hug.API = None)
+```
+
+Which was used in the following manner:
+
+```python
+@hug.get()
+def my_endpoint(hug_timer):
+    pass
+
+
+# OR in the more modern form:
+
+
+@hug.get()
+def my_endpoint(timer: hug.directives.timer):
+    pass
+```
+
+
+### Defining a dependency
+
+Defining a dependency will happen via a new hug `provides` decorator:
+
+```python
+    @hug.provides()
+    def mysql_connection(request, config):
+        ...
```

Comment:
I find myself slightly confused by this example. Is `request` a magic keyword talking about a request object, or is it just the name of a `?request=` or `--request` parameter from the caller? If the latter I'd recommend instead providing an example with `host` and `port` or `user` and `db` because those are more explicitly mysql configuration items.

Reply:
Agreed :), I've updated to be host and port. request was a magic variable, however all magic variables should be replaced in Hug 3.0.0 with this dependency system. So if a user needed the request:

```
@hug.provides()
def mysql_connection(request=requires("request")):
     ....
```
would be used instead.

Reply:
Makes sense, but we should have a list of reserved `provides` keywords per interface. Or we can just rely on the `override=False` system to warn the user.